x/sequencer/keeper/rotation.go

```diff
@@ -0,0 +1,136 @@
+		sdk.NewEvent(
+			types.EventTypeNoticePeriodStarted,
+			sdk.NewAttribute(types.AttributeKeyRollappId, seq.RollappId),
+			sdk.NewAttribute(types.AttributeKeySequencer, seq.SequencerAddress),
+			sdk.NewAttribute(types.AttributeKeyNextProposer, nextSeq.SequencerAddress),
+			sdk.NewAttribute(types.AttributeKeyCompletionTime, completionTime.String()),
+		),
+	)
+
+	return completionTime
+}
+
+// MatureSequencersWithNoticePeriod start rotation flow for all sequencers that have finished their notice period
+// The next proposer is set to the next bonded sequencer
+// The hub will expect a "last state update" from the sequencer to start unbonding
+func (k Keeper) MatureSequencersWithNoticePeriod(ctx sdk.Context, currTime time.Time) {
+	seqs := k.GetMatureNoticePeriodSequencers(ctx, currTime)
+	for _, seq := range seqs {
+		if k.IsProposer(ctx, seq.RollappId, seq.SequencerAddress) {
+			k.StartRotation(ctx, seq.RollappId)
+			k.removeNoticePeriodSequencer(ctx, seq)
+		}
+	}
+}
+
+// IsRotating returns true if the rollapp is currently rotating proposers
+func (k Keeper) IsRotating(ctx sdk.Context, rollappId string) bool {
+	return k.HasNextProposer(ctx, rollappId)
+}
+
+// IsNoticePeriodRequired returns true if the sequencer requires a notice period before unbonding
+// Both the proposer and the next proposer require a notice period
+func (k Keeper) IsNoticePeriodRequired(ctx sdk.Context, seq types.Sequencer) bool {
+	return k.IsProposer(ctx, seq.RollappId, seq.SequencerAddress) || k.IsNextProposer(ctx, seq.RollappId, seq.SequencerAddress)
+}
+
+// ExpectedNextProposer returns the next proposer for a rollapp
+func (k Keeper) ExpectedNextProposer(ctx sdk.Context, rollappId string) types.Sequencer {
+	// if nextProposer is set, were in the middle of rotation
+	seq, ok := k.GetNextProposer(ctx, rollappId)
+	if ok {
+		return seq
+	}
+
+	seqs := k.GetSequencersByRollappByStatus(ctx, rollappId, types.Bonded)
+	if len(seqs) == 0 {
+		return types.Sequencer{}
+	}
+
+	// take the next bonded sequencer to be the proposer. sorted by bond
+	sort.SliceStable(seqs, func(i, j int) bool {
+		return seqs[i].Tokens.IsAllGT(seqs[j].Tokens)
+	})
+
+	// return the first sequencer that is not the proposer
+	active, _ := k.GetProposer(ctx, rollappId)
+	for _, s := range seqs {
+		if s.SequencerAddress != active.SequencerAddress {
+			return s
+		}
+	}
+
+	return types.Sequencer{}
+}
+
+// StartRotation sets the nextSequencer for the rollapp.
+// This function will not clear the current proposer
+// This function called when the sequencer has finished its notice period
+func (k Keeper) StartRotation(ctx sdk.Context, rollappId string) {
+	// next proposer can be empty if there are no bonded sequencers available
+	nextProposer := k.ExpectedNextProposer(ctx, rollappId)
+	k.SetNextProposer(ctx, rollappId, nextProposer.SequencerAddress)
+
+	k.Logger(ctx).Info("rotation started", "rollappId", rollappId, "nextProposer", nextProposer.SequencerAddress)
```

Comment:
it seems a bit misleading as if there is no next proposer, you still log this message and report rotation (though there is no rotation)

Reply:
there is rotation. to empty next proposer.
not sure we need to add additional terminology.
we can call it "handover" instead of rotation